File: 0115-distinct-subsequences/0115-distinct-subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        ns,nt=len(s),len(t)
        #dp = [x] * nt
        dp = [0] * (nt+1)
        dp[0] = 1
        for c in s:
            for i in range(nt,0,-1):
                dp[i] = dp[i] + (t[i-1]==c) * dp[i-1]
        return dp[nt]
        

File: 0115-distinct-subsequences/0115-distinct-subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        int ns = s.length(), nt = t.length();
        vector<int> ind[256];
        vector<unsigned int> dp(nt+1);
        for(int i=nt-1;i+1;--i) ind[t[i]].push_back(i+1);
        dp[0] = 1;
        for(char c:s) for(int i: ind[c]) dp[i] += dp[i-1];
        return dp[nt];
    }
};